File: ROS/src/control/pathplanning/pkgs/triangulation/paths.py
import numpy as np
from pathplanning_dc.node import Node
import triangulation.utils as utils
from typing import Tuple
import math


class TriangulationPaths:
    def __init__(
        self,
        max_iter: int,
        max_angle_change: float,
        max_path_distance: float,
        safety_dist: float,
        continuous_dist: float,
        stage1_rectangle_width: float,
        stage1_bad_points_threshold: int,
        stage1_center_points_threshold: int,
        stage2_rectangle_width: float,
        stage2_bad_points_threshold: int,
        stage2_center_points_threshold: int,
        max_depth: int,
    ) -> None:
        """Initialize triangulator

        Args:
            max_iter: Amount of iterations
            max_angle_change: Maximum angle change in path
            max_path_distance: Maximum distance between nodes in the planned path
            safety_dist: Safety distance from objects
            continuous_dist: the max distance between nodes in stage 1
            stage1_rect_width: width of the rectangle for line expansion
            stage1_bad_points_threshold: threshold for the amount of bad points crossings allowed
            stage1_center_points_threshold: threshold for the amount of center point (not used) crossings allowed
            stage2_rect_width: width of the rectangle for line expansion
            stage2_bad_points_threshold: threshold for the amount of bad points crossings allowed
            stage2_center_points_threshold: threshold for the amount of center point (not used) crossings allowed
            max_depth: maximal depth for path searching in stage 2
        """
        self.max_iter = max_iter
        self.max_angle_change = max_angle_change
        self.max_path_distance = max_path_distance
        self.safety_dist = safety_dist
        self.safety_dist_squared = safety_dist**2

        self.continuous_dist = continuous_dist
        self.stage1_rectangle_width = stage1_rectangle_width
        self.stage1_bad_points_threshold = stage1_bad_points_threshold
        self.stage1_center_points_threshold = stage1_center_points_threshold
        self.stage2_rectangle_width = stage2_rectangle_width
        self.stage2_bad_points_threshold = stage2_bad_points_threshold
        self.stage2_center_points_threshold = stage2_center_points_threshold
        self.max_depth = max_depth

    def get_all_paths(
        self,
        bad_points: np.ndarray,
        center_points: np.ndarray,
        cones: np.ndarray,
        range_front: float,
    ) -> Tuple[np.ndarray, np.ndarray]:
        """Get/generate all possible paths

        Args:
            triangulation_centers: center points of the edges of the triangles
            center_points: center points of the edges of the triangles that were duplicated,
                probably forming the center line
            cones: cones
            range_front: The lookahead distance for sorting points

        Returns:
        tuple of root node and leaves of paths
        """
        root_node = Node(0, 0, 0, None, [], 0, 0)
        queue = [(root_node, [(0, 0)], 0)]
        leaves = []

        iterations = 0
        while queue:
            iterations += 1

            if iterations >= self.max_iter:
                break

            parent, path, depth = queue.pop()

            # First stage is adding as much continuous nodes as possible to reduce search space by alot
            child_found = True
            while child_found:

                next_nodes = utils.sort_closest_to(
                    center_points, (parent.x, parent.y), self.continuous_dist
                )
                child_found = False
                for next_pos in next_nodes:

                    if utils.check_if_feasible_child(
                        parent,
                        path,
                        next_pos,
                        bad_points,
                        center_points,
                        cones,
                        self.max_angle_change,
                        self.safety_dist_squared,
                        self.stage1_rectangle_width,
                        self.stage1_bad_points_threshold,
                        self.stage1_center_points_threshold,
                    ):

                        distance_node = (parent.x - next_pos[0]) ** 2 + (
                            parent.y - next_pos[1]
                        ) ** 2
                        angle_node = np.arctan2(
                            next_pos[1] - parent.y, next_pos[0] - parent.x
                        )
                        angle_change = angle_node - parent.angle

                        # All good! Add this one
                        node = Node(
                            next_pos[0],
                            next_pos[1],
                            distance_node,
                            parent,
                            [],
                            angle_node,
                            angle_change,
                        )

                        # Move node
                        path += [(node.x, node.y)]
                        parent.children.append(node)
                        parent = parent.children[0]

                        child_found = True

                        # A single child is enough!
                        break

            # Now we are at the second stage. We basically look for a node in the distance (to bridge a gap) and add it to the stack to explore further
            # We also limit the depth of this stage
            if depth >= self.max_depth:
                leaves.append(parent)
                continue

            next_nodes = utils.sort_closest_to(
                center_points, (parent.x, parent.y), range_front
            )
            child_found = False
            for next_pos in next_nodes:

                if utils.check_if_feasible_child(
                    parent,
                    path,
                    next_pos,
                    bad_points,
                    center_points,
                    cones,
                    self.max_angle_change,
                    self.safety_dist_squared,
                    self.stage2_rectangle_width,
                    self.stage2_bad_points_threshold,
                    self.stage2_center_points_threshold,
                ):

                    distance_node = (parent.x - next_pos[0]) ** 2 + (
                        parent.y - next_pos[1]
                    ) ** 2
                    angle_node = np.arctan2(
                        next_pos[1] - parent.y, next_pos[0] - parent.x
                    )
                    angle_change = angle_node - parent.angle

                    # All good! Add this one
                    node = Node(
                        next_pos[0],
                        next_pos[1],
                        distance_node,
                        parent,
                        [],
                        angle_node,
                        angle_change,
                    )

                    # Move node
                    new_path = path + [(node.x, node.y)]
                    parent.children.append(node)

                    queue.append((node, new_path, depth + 1))

                    child_found = True

            # If no (valid) child can be found it means that we have found a leaf
            if not child_found:
                leaves.append(parent)

        return None, leaves

    def get_cost_branch(
        self,
        branch: np.ndarray,
        cones: np.ndarray,
        range_front: float,
    ) -> Tuple[float, float]:
        """Get cost of branch.

        Args:
            branch: branch to calculate cost for
            cones: cones
            range_front: The lookahead distance for sorting points

        Returns:
        tuple of angle_cost, color_cost, width_cost, spacing_cost, length_cost
        """
        # angle should not change much over one path
        angle_changes = np.array([abs(point.angle_change) for point in branch])
        angle_cost = np.var(angle_changes)

        # longer paths usually work better as they make use of more center points
        # having many segments along a path is usually a good path
        node_distances = np.array([point.distance for point in branch])
        distance = np.sum(node_distances)
        length_cost = 1 / distance + 1 / len(node_distances) * 10

        # get array of blue cones and array of yellow cones sorted by distance
        sorted_cones = utils.sort_closest_to(cones, (0, 0), range_front)
        blue_sorted = sorted_cones[sorted_cones[:, 2] == 0]
        yellow_sorted = sorted_cones[sorted_cones[:, 2] == 1]

        # Iterate over the path, get all cones in a certain distance to the line, and apply a penalty every time a cone is on the wrong side
        for point in branch:
            line_distances_squared = np.cos(point.angle) * (
                cones[:, 1] - point.y
            ) - np.sin(point.angle) * (cones[:, 0] - point.x)
            distances_squared = utils.distance_squared(
                point.x, point.y, cones[:, 0], cones[:, 1]
            )
            close_cones = cones[
                np.logical_and(
                    line_distances_squared < 3**2, distances_squared < 3**2
                )
            ]
            close_classes = close_cones[:, -1]

            diff_angles = (
                np.arctan2(close_cones[:, 1] - point.y, close_cones[:, 0] - point.x)
                - point.angle
            )

            # Left (positive angle) should be class 0, right (negative angle) should be class 1
            # Ignore class >= 2
            penalty_amount = np.sum(close_classes[diff_angles > 0.0] == 1) + np.sum(
                close_classes[diff_angles < 0.0] == 0
            )

            length_cost += penalty_amount * 10
            angle_cost += penalty_amount * 10

        # get center between closest blue cone and closest yellow cone
        center_point = (
            (blue_sorted[0][0] + yellow_sorted[0][0]) / 2,
            (blue_sorted[0][1] + yellow_sorted[0][1]) / 2,
        )
        for point in branch:
            if np.isclose(center_point[0], point.x) and np.isclose(
                center_point[1], point.y
            ):
                # if any point of the path is close to this center point, reduce cost
                length_cost /= 100
                angle_cost /= 100

        return angle_cost, length_cost

File: ROS/src/control/pathplanning/src/paths.cpp
#include <paths.hpp>
#include <ros/ros.h>

namespace pathplanning {
std::pair<Node *, std::vector<Node *>> TriangulationPaths::get_all_paths(
    const std::vector<std::vector<double>> &bad_points,
    const std::vector<std::vector<double>> &center_points,
    const std::vector<std::vector<double>> &cones) {
  Node *root_node = new Node(0, 0, 0, nullptr, std::vector<Node *>(), 0, 0);

  // Queue contains node to search, path already searched (corresponding to that
  // node), current depth
  std::vector<std::tuple<Node *, std::vector<std::array<double, 2>>, int>>
      queue({std::make_tuple(root_node,
                             std::vector<std::array<double, 2>>(
                                 {std::array<double, 2>({0.0, 0.0})}),
                             0)});

  std::vector<Node *> leaves;
  int iteration = 0;

  while (!queue.empty()) {
    iteration++;

    if (iteration >= this->max_iter)
      break;

    // Get the next element from the queue
    auto t = queue[0];
    auto parent = std::get<0>(t);
    auto path = std::get<1>(t);
    auto depth = std::get<2>(t);

    queue.erase(queue.begin());

    bool first_pose_found = true;
    if (path.size() <= 1) {
      first_pose_found = false;
    }

    bool child_found = false;

    // Handle first pose
    if (!first_pose_found) {
      bool second_child_found = false;

      // Search all nodes in continuous distance and add node with smallest
      // angle_change from car orientation
      double min_angle_change = 1000;
      Node *node = nullptr;
      Node *node2 = nullptr;

      // Get the center points to this element that are within close
      // distance and sort them by angle_change
      std::vector<PointInfo> next_nodes;
      next_nodes = sort_by_angle_change(center_points, {parent->x, parent->y},
                                        parent->angle, this->max_angle_change,
                                        this->continuous_dist_);

      for (PointInfo next_node : next_nodes) {
        if (pathplanning::check_if_feasible_child(
                *parent, path, next_node.point, bad_points, center_points,
                cones, this->max_angle_change, this->safety_dist_squared, 0,
                this->stage1_threshold_bad_points_,
                this->stage1_threshold_center_points_)) {

          if (!child_found) {
            node = new Node(next_node.point[0], next_node.point[1],
                            next_node.distance_squared, parent,
                            std::vector<Node *>(), next_node.angle,
                            next_node.angle_change);
            child_found = true;
          } else {
            if (!second_child_found) {
              if (node->x == next_node.point[0] &&
                  node->y == next_node.point[1]) {
                continue;
              }
              node2 = new Node(next_node.point[0], next_node.point[1],
                               next_node.distance_squared, parent,
                               std::vector<Node *>(), next_node.angle,
                               next_node.angle_change);
              second_child_found = true;
              break;
            }
          }
        }
      }
      if (child_found) {
        if (second_child_found) {
          std::vector<std::array<double, 2>> new_path(path);
          new_path.push_back({node2->x, node2->y});
          parent->children.push_back(node2);

          queue.push_back(std::make_tuple(node2, new_path, depth + 1));
        }

        path.push_back({node->x, node->y});
        parent->children.push_back(node);
        parent = node;
      }
    }

    // Get the closest center points to this element that are within close
    // distance (stage 1)
    std::vector<std::vector<double>> next_nodes;
    next_nodes = sort_closest_to(center_points, {parent->x, parent->y},
                                 this->continuous_dist_);

    // if first point of path close to last point of path, stop loop
    if (path.size() > 20 &&
        distance_squared(path[1][0], path[1][1], next_nodes[0][0],
                         next_nodes[0][1]) < this->close_path_dist) {
      leaves.push_back(parent);
      break;
    }

    child_found = false;

    float max_dist = pow(this->continuous_dist_, 2);

    for (auto next_pos : next_nodes) {
      if (pathplanning::check_if_feasible_child(
              *parent, path, next_pos, bad_points, center_points, cones,
              this->stage1_max_angle_change_, this->safety_dist_squared,
              this->stage1_rect_width_, this->stage1_threshold_bad_points_,
              this->stage1_threshold_center_points_)) {

        double distance_node =
            pow(parent->x - next_pos[0], 2) + pow(parent->y - next_pos[1], 2);
        double angle_node =
            atan2(next_pos[1] - parent->y, next_pos[0] - parent->x);
        double angle_change = angle_node - parent->angle;

        Node *node = new Node(next_pos[0], next_pos[1], distance_node, parent,
                              std::vector<Node *>(), angle_node, angle_change);

        if (distance_node > max_dist) {
          break;
        }

        std::vector<std::array<double, 2>> new_path(path);
        new_path.push_back({node->x, node->y});
        parent->children.push_back(node);

        queue.push_back(std::make_tuple(node, new_path, depth + 1));

        if (!child_found) {
          max_dist = distance_node + this->stage1_dist_window_;
          child_found = true;
        }
      }
    }

    if (child_found) {
      continue;
    }

    // Now we are at the second stage. We basically look for a node in the
    // distance (to bridge a gap) and add it to the stack to explore further We
    // also limit the depth of this stage (stage 2)
    if (depth >= this->max_depth_) {
      leaves.push_back(parent);
      continue;
    }

    // Get the closest center points to this element that are within
    // stage2_max_dist
    std::vector<std::vector<double>> next_nodes2;
    next_nodes2 = sort_closest_to(center_points, {parent->x, parent->y},
                                  this->stage2_max_dist_);
    child_found = false;

    for (auto next_pos : next_nodes2) {
      if (pathplanning::check_if_feasible_child(
              *parent, path, next_pos, bad_points, center_points, cones,
              this->max_angle_change, this->safety_dist_squared,
              this->stage2_rect_width_, this->stage2_threshold_bad_points_,
              this->stage2_threshold_center_points_)) {

        double distance_node =
            pow(parent->x - next_pos[0], 2) + pow(parent->y - next_pos[1], 2);
        double angle_node =
            atan2(next_pos[1] - parent->y, next_pos[0] - parent->x);
        double angle_change = angle_node - parent->angle;

        // The first pose of the path is an exeption and its angle should be
        // oriented in the same direction as the car
        if (!first_pose_found) {
          angle_node = 0;
        }

        Node *node = new Node(next_pos[0], next_pos[1], distance_node, parent,
                              std::vector<Node *>(), angle_node, angle_change);

        std::vector<std::array<double, 2>> new_path(path);
        new_path.push_back({node->x, node->y});
        parent->children.push_back(node);

        queue.push_back(std::make_tuple(node, new_path, depth + 1));
      }
    }

    if (!child_found) {
      leaves.push_back(parent);
    }
  }

  return std::make_pair(root_node, leaves);
}

std::pair<double, double> TriangulationPaths::get_cost_branch(
    const std::vector<Node *> &branch,
    const std::vector<std::vector<double>> &cones) {

  std::vector<double> angle_changes;
  std::vector<double> node_distances;
  for (const Node *point : branch) {
    angle_changes.push_back(std::abs(point->angle_change));
    node_distances.push_back(point->distance);
  }

  // angle should not change much over one path
  double angle_cost = calculate_variance(angle_changes);

  // longer paths usually work better as they make use of more center points
  // having many segments along a path is usually a good path
  double distance =
      std::accumulate(node_distances.begin(), node_distances.end(), 0.0);
  double length_cost = 1 / distance + 10.0 / node_distances.size();

  double penalty = 0;

  // Iterate over the path, get all cones in a certain distance to line, and
  // apply a penalty every time a cone is on the wrong side
  for (const auto &point : branch) {
    std::vector<double> line_distances_squared(cones.size());
    std::vector<double> distances_squared(cones.size());
    std::vector<std::vector<double>> close_cones;
    std::vector<double> diff_angles;
    std::vector<int> close_classes;

    for (size_t i = 0; i < cones.size(); ++i) {
      line_distances_squared[i] =
          std::cos(point->angle) * (cones[i][1] - point->y) -
          std::sin(point->angle) * (cones[i][0] - point->x);
      distances_squared[i] =
          distance_squared(point->x, point->y, cones[i][0], cones[i][1]);

      if (line_distances_squared[i] < 5 * 3 && distances_squared[i] < 5 * 3) {
        close_cones.push_back(cones[i]);
        diff_angles.push_back(
            std::atan2(cones[i][1] - point->y, cones[i][0] - point->x) -
            point->angle);
        if (diff_angles.back() > M_PI) {
          diff_angles.back() -= 2 * M_PI;
        } else if (diff_angles.back() < -M_PI) {
          diff_angles.back() += 2 * M_PI;
        }
        close_classes.push_back(cones[i][2]);
      }
    }

    double penalty_amount = 0;
    for (size_t i = 0; i < close_cones.size(); ++i) {
      if (diff_angles[i] > 0.0 && close_classes[i] == 1) {
        penalty_amount++;
      } else if (diff_angles[i] < 0.0 && close_classes[i] == 0) {
        penalty_amount++;
      }
    }

    penalty += penalty_amount / 100;
  }

  return std::make_pair(angle_cost + penalty, length_cost + penalty);
}

} // namespace pathplanning